### PR TITLE
UseInViewSelect

### PR DESCRIPTION
**Pull request by zutigrm**
## Summary

Addresses issue:

- #8789

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/hooks/useInViewSelect.js [hunk 1]
```diff
@@ -40,15 +40,17 @@ import { useInView } from './useInView';
 export const useInViewSelect = ( mapSelect, deps = [] ) => {
 	const isInView = useInView( { sticky: true } );
 	const latestSelectorResult = useRef();
+	const notInViewCallback = () => {
```
Comment (eugene-manuilov) [inline assets/js/hooks/useInViewSelect.js:43]: > ... to a stable named function outside of the component.

Let's move this out of the `useInViewSelect` hook.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @zutigrm. Mostly looks good to me. Left one comment.

File: assets/js/hooks/useInViewSelect.js [hunk 2]
```diff
@@ -27,6 +27,10 @@ import { useCallback, useRef } from '@wordpress/element';
  */
 import { useInView } from './useInView';
 
+const notInViewCallback = () => {
+	return undefined;
+};
```
Comment (eugene-manuilov) [inline assets/js/hooks/useInViewSelect.js:32]: Are there specific reasons why it must be an arrow function? We should use the standard function declaration whenever possible.

```suggestion
function notInViewCallback() {
	return undefined;
}
```

**Comment by zutigrm** [comment]
@eugene-manuilov Thanks, I updated PR. I noticed that few key metrics widgets were stuck in loading state, so after digging up a bit, I also needed to update the usage of this hook, as we tend to use it for fetching reports, so I updated the usage to include proper dependencies - beside the initial resolution of some dependencies, report options can update when reference date is changed on the dashboard, so this PR grew a bit in number of files that needed editing.

File: assets/js/hooks/useInViewSelect.js (hunk 2 above)
Comment (zutigrm) [reply assets/js/hooks/useInViewSelect.js]: Got it, changed.

Comment (zutigrm) [reply assets/js/hooks/useInViewSelect.js]: Got it, changed.

File: assets/js/hooks/useInViewSelect.js [hunk 3]
```diff
@@ -27,6 +27,10 @@ import { useCallback, useRef } from '@wordpress/element';
  */
 import { useInView } from './useInView';
 
+function notInViewCallback() {
```
Comment (eugene-manuilov) [inline assets/js/hooks/useInViewSelect.js:30]: Forgot to ask you to add jsdoc to this function.

File: assets/js/modules/analytics-4/components/widgets/LeastEngagingPagesWidget.js [hunk 4]
```diff
@@ -137,13 +142,15 @@ function LeastEngagingPagesWidget( props ) {
...
-			: undefined
+	const titles = useInViewSelect(
+		( select ) =>
+			! error
+				? select( MODULES_ANALYTICS_4 ).getPageTitles(
+						report,
+						reportOptions
+				  )
+				: undefined,
+		[ report, reportOptions ]
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/components/widgets/LeastEngagingPagesWidget.js:153]: We need to add `error` as dependency as well.

**Review by eugene-manuilov — COMMENTED** [review]
Great job, @zutigrm. Almost looks good to me. Left two minor comments for you and then it should be good to go.

File: assets/js/modules/analytics-4/components/widgets/LeastEngagingPagesWidget.js (hunk 4 above)
Comment (zutigrm) [reply assets/js/modules/analytics-4/components/widgets/LeastEngagingPagesWidget.js:153]: Great spot, thanks, I missed that one

**Comment by zutigrm** [comment]
Thanks @eugene-manuilov , PR updated

**Review by eugene-manuilov — APPROVED** [review]
Thanks, @zutigrm. LGTM.

**Comment by aaemnnosttv** [comment]
> I noticed that few key metrics widgets were stuck in loading state, so after digging up a bit, I also needed to update the usage of this hook, as we tend to use it for fetching reports, so I updated the usage to include proper dependencies - beside the initial resolution of some dependencies, report options can update when reference date is changed on the dashboard, so this PR grew a bit in number of files that needed editing.

Thanks @zutigrm – while I'm not opposed to the addition of dependencies in the usage, it shouldn't have been necessary here. This is likely due to a related flaw in the initial implementation which set `deps = []` by default. The inclusion of the `mapSelect` in `deps` which this issue was intended to fix prevented this from being a problem since `mapSelect` was always different. Removing it caused our `deps` to use the default empty array in any instance where we weren't passing deps explicitly. This resulted in the selector callback never updating when its dependencies changed. Again, it's good to address, but it ultimately resulted in a much larger change than was necessary.

I've opened #9009 to address the fix for the default.

**Pull request merged**